### PR TITLE
Account Transaction : Update Contract

### PR DESCRIPTION
**Pull request by parv0888**
## Purpose
adds support for account transaction : update contract

## Changes
adds support for account transaction : update contract

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

## CLA acceptance

By submitting the contribution I accept the terms and conditions of the
Contributor License Agreement v1.0
- link: https://developers.concordium.com/CLAs/Contributor-License-Agreement-v1.0.pdf

- [x] I accept the above linked CLA.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -777,6 +777,43 @@ try{
...
+ ```java
+try{
+    byte[] paramBytes = new byte[0];
+    TransactionSigner signer = TransactionSigner.from(
+        SignerEntry.from(Index.from(0), Index.from(0),
+            firstSecretKey),
+        SignerEntry.from(Index.from(0), Index.from(1),
+            secondSecretKey));
+
+    val transaction = TransactionFactory.newUpdateContract()
```
Comment (EmilLa1) [inline README.md:792]: I don't think the README/docs should be using `Lombok` features (like `val`).

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/Payload.java [hunk 2]
```diff
@@ -91,7 +91,9 @@ private static UInt64 calculateEnergyCost(int noOfSignatures,
     public enum PayloadType {
         TRANSFER,
         TRANSFER_WITH_MEMO,
-        REGISTER_DATA
+        REGISTER_DATA,
+        UPDATE_SMART_CONTRACT_INSTANCE,
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/Payload.java:95]: Please just name it `Update` so it aligns with `concordium-base`. I think that will make it easier for future maintainers.

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/ReceiveName.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * A receive name (owned version). Expected format:
+ * "<contract_name>.<method_name>".
+ */
+
+@Getter
+@ToString
+public final class ReceiveName {
+    private final String contractName;
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/ReceiveName.java:20]: Document fields.

File: concordium-sdk/src/test/java/com/concordium/sdk/transactions/UpdateContractPayloadTest.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    @Test
+    public void testCreatePayload() {
+        byte[] emptyArray = new byte[0];
+        val payload = UpdateContractPayload.from(
+                0,
+                ContractAddress.from(81, 0),
+                "CIS2-NFT",
+                "mint",
+                emptyArray);
+        assertNotNull(payload);
```
Comment (EmilLa1) [inline concordium-sdk/src/test/java/com/concordium/sdk/transactions/UpdateContractPayloadTest.java:20]: Maybe assert that `getBytes` returns the expected bytes.

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateContractTransaction.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+import java.util.Objects;
+
+/**
+ * Construct a transaction to update a smart contract instance.
+ */
+@Getter
+public class UpdateContractTransaction extends AbstractTransaction {
+
+    private final UpdateContractPayload payload;
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateContractTransaction.java:18]: I'm aware it's trivial for you what the fields are meant for. 
But please add a bit of documentation so future users can easily lookup what the fields mean/are used for.

**Review by EmilLa1 — CHANGES_REQUESTED** [review]
Looks good!
Few minor comments mainly regarding naming and some missing documentation.

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/ReceiveName.java (hunk 3 above)
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/transactions/ReceiveName.java:23]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/Payload.java (hunk 2 above)
Comment (tds-1) [reply concordium-sdk/src/main/java/com/concordium/sdk/transactions/Payload.java]: Done

File: README.md (hunk 1 above)
Comment (tds-1) [reply README.md]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/Parameter.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import lombok.Getter;
+import lombok.ToString;
+import lombok.val;
+
+import java.nio.ByteBuffer;
+
+@Getter
+@ToString
+public final class Parameter {
```
[the remark this answers is not in the document]
Comment (tds-1) [reply concordium-sdk/src/main/java/com/concordium/sdk/transactions/Parameter.java:18]: Done

File: CHANGELOG.md [hunk 7]
```diff
@@ -37,6 +37,7 @@
 - Added support for query `UnBanNode`
 - Added support for query `StartBaker`
 - Added support for query `StopBaker`
+- Added support for Update Contract.
```
[the remark this answers is not in the document]
Comment (tds-1) [reply CHANGELOG.md]: Done

File: concordium-sdk/src/test/java/com/concordium/sdk/transactions/UpdateContractPayloadTest.java (hunk 4 above)
Comment (tds-1) [reply concordium-sdk/src/test/java/com/concordium/sdk/transactions/UpdateContractPayloadTest.java]: Changed

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/Parameter.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (EmilLa1) [reply concordium-sdk/src/main/java/com/concordium/sdk/transactions/Parameter.java:18]: I think you can elaborate a bit here. 

In particular the `Parameters` are the parameters used for `updating` a smart contract instance, i.e. calling a `receive` function exposed in the smart contract.

Further there are some limits on the size of the parameters based given on the protocol version. 
In the current supported protocols it is limited to be of size 1kb. 

(WIth protocol 5 this becomes 65kb, but I will update that in the P5 pull request: https://github.com/Concordium/concordium-java-sdk/pull/144

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateContractPayload.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
+package com.concordium.sdk.transactions;
+
+import lombok.Getter;
+import lombok.ToString;
+import lombok.val;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Data needed to update a smart contract.
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateContractPayload.java:10]:
```suggestion
 * The payload for updating a smart contract.
```

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateContractPayload.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+ */
+@ToString
+@Getter
+public final class UpdateContractPayload {
+    /**
+     * Send the given amount of CCD.
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateContractPayload.java:16]:
```suggestion
     * Send the given amount of CCD to the smart contract.
```

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateContractPayload.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+     */
+    private final CCDAmount amount;
+    /**
+     * Address of the contract instance to invoke.
+     */
+    private final ContractAddress contract_address;
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateContractPayload.java:22]: Please use camelCase

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateContractPayload.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    /**
+     * Name of the method to invoke on the contract.
+     */
+    private final ReceiveName receive_name;
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateContractPayload.java:26]: Please use camelCase.

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateContractTransaction.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+     * Data needed to update a smart contract.
+     */
+    private final UpdateContractPayload payload;
+
+    /**
+     * Account Address of the sender
+     */
+    private final AccountAddress sender;
+    /**
+     * The nonce value/
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateContractTransaction.java:27]:
```suggestion
     * The senders account next available nonce.
```

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateContractTransaction.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    /**
+     * Indicates when the transaction should expire.
+     */
+    private final Expiry expiry;
+    /**
+     * A signer object that is used to sign the transaction.
+     */
+    private final TransactionSigner signer;
+    /**
+     * Maximum energy required for the transaction.
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateContractTransaction.java:39]:
```suggestion
     * Maximum energy **allowed** for the transaction to use.
```

**Review by EmilLa1 — CHANGES_REQUESTED** [review]
Looks good. A few minor things left and it's good to go.

**Comment by EmilLa1** [comment]
Closing as this is part of https://github.com/Concordium/concordium-java-sdk/pull/146

**Pull request closed without merging**